lib/THCUNN/LogSoftMax.cu

```diff
@@ -256,36 +331,100 @@ void THNN_CudaLogSoftMax_updateOutput(THCState *state, THCudaTensor *input, THCu
 {
   THCUNN_assertSameGPU(state, 2, input, output);
 
-  input = THCudaTensor_newContiguous(state, input);
   THCudaTensor_resizeAs(state, output, input);
 
+  bool spatial  = false;
   int batchSize = 1;
   int classSize = 0;
+  int height = 0;
+  int width = 0;
+
+  int ndims = THCudaTensor_nDimension(state, input);
 
-  if (THCudaTensor_nDimension(state, input) == 1)
+  if (ndims == 1)
+  {
+    classSize = THCudaTensor_size(state, input, 0);
+    input = THCudaTensor_newContiguous(state, input);
+  }
+  else if (ndims == 2)
   {
+    batchSize = THCudaTensor_size(state, input, 0);
+    classSize = THCudaTensor_size(state, input, 1);
+    input = THCudaTensor_newContiguous(state, input);
+  }
+  else if (ndims == 3)
+  {
+    spatial = true;
     classSize = THCudaTensor_size(state, input, 0);
+    height = THCudaTensor_size(state, input, 1);
+    width = THCudaTensor_size(state, input, 2);
+
+    THCudaTensor *transposedInput = THCudaTensor_new(state);
+    THCudaTensor_resizeAs(state, transposedInput, input);
+    THCudaTensor_copy(state, transposedInput, input);
+    input = transposedInput;
+    // convert tensor from torch layout to cuda layout
+    // P x H x W -> W x H x P
+    THCudaTensor_transpose(state, input, input, 0, 2);
+    // W x H x P -> H x W x P
+    THCudaTensor_transpose(state, input, input, 0, 1);
+    input = THCudaTensor_newContiguous(state, input);
```

Comment:
Rather than doing a clone of input + transpose + contiguous, which executes two copies, it would be better to create `transposedInput` already with correct sizes and copy `input` into it (copy only cares about the number of elements, not about the dimension sizes). Same goes for transpose in `updateGradInput`.